docs/src/main/asciidoc/qute.adoc

```diff
@@ -0,0 +1,271 @@
+<2> `Template.render()` returns a new rendering action that can be customized before the actual rendering is triggered. In this case, we put the name value under the key `name`. The data map is accessible during rendering. 
+<3> Note that we don't trigger the rendering - this is done automatically by a special `ContainerResponseFilter` implementation.
+
+If running your application, you can request the endpoint:
+
+```
+$ curl -w "\n" http://localhost:8080/hello?name=Martin
+Hello Martin!
+```
+
+== Parameter Declarations and Template Extension Methods
+
+Qute has many useful features.
+In this example, we'll demonstrate two of them.
+If you declare a *parameter declaration* in a template then Qute attempts to validate all expressions that reference this parameter and if an incorrect expression is found the build fails.
+*Template extension methods* are used to extend the set of accessible properties of data objects.
+
+Let's suppose we have a simple class like this:
+
+.Item.java
+[source,java]
+----
+public class Item {
+    public String name;
+    public BigDecimal price;
+}
+----
+
+And we'd like to render a simple HTML page that contains the item name, price and also a discounted price.
+The discounted price is sometimes called a "computed property".
+We will implement a template extension method to render this property easily.
+Let's start again with the template:
+
+.item.html
+[source,html]
+----
+{@org.acme.Item item} <1>
+<!DOCTYPE html>
+<html>
+<head>
+<meta charset="UTF-8">
+<title>{item.name}</title> <2>
+</head>
+<body>
+    <h1>{item.name}</h1>
+    <div>Price: {item.price}</div> 
+    {#if item.price > 100} <3>
+    <div>Discounted Price: {item.discountedPrice}</div> <4>
+    {/if}
+</body>
+</html>
+----
+<1> Optional parameter declaration. Qute attempts to validate all expressions that reference paramater `item`.
+<2> This expression is validated. Try to change the expression to `{item.nonSense}` and the build should fail.
+<3> `if` is a basic control flow section.
+<4> This expression is also validated against the `Item` class and obviously there is no such property declared. However, there is a template extension method declared on the `ItemResource` class - see below.
+
+Finally, let's create a resource class.
+
+.ItemResource.java
+[source,java]
+----
+package org.acme.quarkus.sample;
+
+import javax.inject.Inject;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+
+import io.quarkus.qute.Rendering;
+import io.quarkus.qute.Template;
+
+@Path("item")
+public class ItemResource {
+
+    @Inject
+    ItemService service;
+
+    @Inject
+    Template item; <1>
+    
+    @GET
+    @Path("{id}")
+    @Produces(MediaType.TEXT_HTML)
+    public Rendering get(@PathParam("id") Integer id) {
+        return item.render().data("item", service.findItem(id)); <2>
+    }
+
+    @TemplateExtension <3>
```

Comment:
I think we should advise people from the start to put their extensions in a separate file, by convention, even though they can put them anywhere. Something like `QuteExtensions.java` or `TemplateExtensions.java`.

Reply:
Well, that's a matter of taste. For demos and guides it's definitely easier to declate the method where it's needed. Of course, we should introduce some good practices in the reference guide.  I can also imagine that that tooling would notify the developer that his application already contains a duplicit extension method for example.

Reply:
For a demo, sure. But as soon as you have extension methods in two classes it becomes hard to remember the list and if you're not reimplementing one that already exists.

In fact, in Play 1, the framework had its own `JavaExtensions` class, and you had to extend it to add new methods. You didn't even need an annotation. Even if we don't go that way, we could allow the annotation on the class to avoid adding it on all its methods, if we advise for grouping.

Reply:
> becomes hard to remember the list and if you're not reimplementing 

Yes, see my comment about tooling above.

> You didn't even need an annotation.

There is `TemplateExtension.matchName()` which could be useful. I'd stay with the current approach for now.

Reply:
> Yes, see my comment about tooling above.

That's a way to generate errors on duplicates, but doesn't help you finding the set of extension methods. If we have a convention it becomes trivial: "Ctrl Shift r > WhateverExtension" and you have the full list ;)

> There is TemplateExtension.matchName() which could be useful. I'd stay with the current approach for now.

 Absolutely, but I mean if we advise to group them in a file, we could allow to place the annotation on the class and all its static methods would be registered even if they didn't have any method-annotation. I'm not saying we must remove the method annotation.